hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/HASecurityUtils.java

```diff
@@ -0,0 +1,266 @@
+            scmStorageConfig.getScmId(), null);
+
+    rootCAServer.init(new SecurityConfig(config),
+        CertificateServer.CAType.SELF_SIGNED_CA);
+
+    try {
+      X509CertificateHolder certificateHolder = rootCAServer.
+          requestCertificate(csr, KERBEROS_TRUSTED, SCM).get();
+
+      X509CertificateHolder rootCAServerCACertificateHolder =
+          rootCAServer.getCACertificate();
+
+      String pemEncodedCert =
+          CertificateCodec.getPEMEncodedString(certificateHolder);
+
+      String pemEncodedRootCert =
+          CertificateCodec.getPEMEncodedString(rootCAServerCACertificateHolder);
+
+
+      client.storeCertificate(pemEncodedRootCert, true, true);
+      client.storeCertificate(pemEncodedCert, true);
+
+      // Write to the location, so that Default CA server can read the
+      // certificate generated by cert client.
+      CertificateCodec certCodec =
+          new CertificateCodec(new SecurityConfig(config),
+              client.getComponentName());
+
+      certCodec.writeCertificate(certificateHolder);
+
+      // Persist scm cert serial ID.
+      scmStorageConfig.setScmCertSerialId(certificateHolder.getSerialNumber()
+          .toString());
+    } catch (InterruptedException e) {
+      LOG.error("Error while storing SCM signed certificate.", e);
+      Thread.currentThread().interrupt();
+      throw new RuntimeException(e);
+    } catch (ExecutionException e) {
+      LOG.error("Error while storing SCM signed certificate.", e);
+      throw new RuntimeException(e);
+    } catch (IOException | CertificateException  e) {
+      LOG.error("Error while storing SCM signed certificate.", e);
+      throw new RuntimeException(e);
+    }
+
+  }
+
+  /**
+   * This function creates/initializes a certificate server as needed.
+   * This function is idempotent, so calling this again and again after the
+   * server is initialized is not a problem.
+   *
+   * @param clusterID - Cluster ID
+   * @param scmID     - SCM ID
+   */
+  public static CertificateServer initializeRootCertificateServer(
+      String clusterID, String scmID, SCMCertStore scmCertStore)
+      throws IOException {
+    String subject = "scm-rootca@" + InetAddress.getLocalHost().getHostName();
+
+    return new DefaultCAServer(subject, clusterID, scmID, scmCertStore,
+        new DefaultCAProfile(),
+        Paths.get(SCM_CA_CERT_STORAGE_DIR, SCM_CA_PATH).toString());
+  }
+
+  private static PKCS10CertificationRequest generateCSR(
+      CertificateClient client, SCMStorageConfig scmStorageConfig,
+      OzoneConfiguration config, InetSocketAddress scmAddress,
+      String fetchedSCMId) throws IOException {
+    CertificateSignRequest.Builder builder = client.getCSRBuilder();
+    KeyPair keyPair = new KeyPair(client.getPublicKey(),
+        client.getPrivateKey());
+
+    // Get host name.
+    String hostname = scmAddress.getAddress().getHostName();
+
+    String subject = "scm@"+ hostname;
```

Comment:
this sub scm cert should have subject like scm-sub@host.

Reply:
Done